homeassistant/components/ring/sensor.py

```diff
@@ -59,183 +62,170 @@ async def async_setup_entry(
-                    found = entry
-                    break
-
-        if not found:
-            return
-
-        self._latest_event = found
-        super()._handle_coordinator_update()
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        if self._latest_event is None:
-            return None
 
-        return self._latest_event["created_at"]
-
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
-        attrs = super().extra_state_attributes
+        self._device = self._get_coordinator_data().get_device(
+            self._device.device_api_id
+        )
+        # History values can drop off the last 10 events so only update
+        # the value if it's not None
+        if native_value := self.entity_description.value_fn(self._device):
+            self._attr_native_value = native_value
+        if extra_attrs := self.entity_description.extra_state_attributes_fn(
+            self._device
+        ):
+            self._attr_extra_state_attributes = extra_attrs
+        super()._handle_coordinator_update()
 
-        if self._latest_event:
-            attrs["created_at"] = self._latest_event["created_at"]
-            attrs["answered"] = self._latest_event["answered"]
-            attrs["recording_status"] = self._latest_event["recording"]["status"]
-            attrs["category"] = self._latest_event["kind"]
 
-        return attrs
+def _get_last_event(
+    history_data: list[dict[str, Any]], kind: RingEventKind | None
+) -> dict[str, Any] | None:
+    if not history_data:
+        return None
+    if kind is None:
+        return history_data[0]
+    for entry in history_data:
+        if entry["kind"] == kind.value:
+            return entry
+    return None
+
+
+def _get_last_event_attrs(
+    history_data: list[dict[str, Any]], kind: RingEventKind | None
+) -> dict[str, Any] | None:
+    if last_event := _get_last_event(history_data, kind):
+        return {
+            "created_at": last_event.get("created_at"),
+            "answered": last_event.get("answered"),
+            "recording_status": last_event.get("recording", {}).get("status"),
+            "category": last_event.get("kind"),
+        }
+    return None
 
 
 @dataclass(frozen=True, kw_only=True)
-class RingSensorEntityDescription(SensorEntityDescription):
+class RingSensorEntityDescription(SensorEntityDescription, Generic[_RingGenericT]):
     """Describes Ring sensor entity."""
 
-    category: list[str]
-    cls: type[RingSensor]
-
-    kind: str | None = None
+    value_fn: Callable[[_RingGenericT], StateType] = lambda _: True
+    exists_fn: Callable[[RingGeneric], bool] = lambda _: True
+    extra_state_attributes_fn: Callable[[_RingGenericT], dict[str, Any] | None] = (
+        lambda _: None
+    )
 
 
 SENSOR_TYPES: tuple[RingSensorEntityDescription, ...] = (
-    RingSensorEntityDescription(
+    RingSensorEntityDescription[RingGeneric](
```

Comment:
```suggestion
    RingSensorEntityDescription(
```
With `default=RingGeneric` now set, you can remove the `RingGeneric` subscript here (and from all other descriptions below.

_It's only necessary in case it's different. E.g. for `RingDoorBell | RingChime` and `RingOther`._